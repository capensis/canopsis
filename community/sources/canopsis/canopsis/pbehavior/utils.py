"""
Utils for pbehaviors.
"""

from dateutil.rrule import rrulestr
from dateutil import tz
from datetime import datetime


EXDATE_DATE_FORMAT = "%Y/%m/%d %H:%M:%S"


def check_valid_rrule(rrule):
    """
    Check for RRULE validity.

    Raises error when invalid with a more explicit message.
    :param str rrule: rrule as string
    :raises ValueError: rrule is invalid
    :returns: True if ok, raises exception if invalid
    :rtype: bool
    """
    if rrule == '' or rrule is None:
        return True

    try:
        rrulestr(rrule)
    except ValueError as ex:
        raise ValueError('Invalid RRULE: {}'.format(ex))

    return True
