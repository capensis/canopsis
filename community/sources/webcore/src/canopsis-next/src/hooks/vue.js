import { getCurrentInstance } from 'vue';

/**
 * Hook for get current component instance
 *
 * @return {Vue}
 */
export const useComponentInstance = () => {
  const instance = getCurrentInstance();

  return instance?.proxy || instance;
};

/**
 * Hook for get current component options
 *
 * @return {Object}
 */
export const useComponentOptions = () => {
  const instance = useComponentInstance();

  return instance.$options;
};

/**
 * Hook for get current component model
 *
 * @return {Object}
 */
export const useComponentModel = () => {
  const { model } = useComponentOptions();

  return {
    event: model?.event ?? 'input',
    prop: model?.event ?? 'value',
  };
};
