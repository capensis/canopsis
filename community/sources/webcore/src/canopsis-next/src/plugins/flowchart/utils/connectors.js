import { range } from 'lodash';

import uid from '@/helpers/uid';

/**
 * @typedef {Object} ConnectorPoint
 * @property {number} x
 * @property {number} y
 * @property {string} _id
 */

/**
 * Generate connector
 *
 * @param {number} x
 * @param {number} y
 * @returns {ConnectorPoint}
 */
export const generateConnectorPoint = (x, y) => ({
  x,
  y,
  _id: uid(),
});

/**
 * Generate connector points by rectangle
 *
 * @param {Rect} rect
 * @param {number} count
 * @returns {ConnectorPoint[]}
 */
export const generateConnectorsByRect = (rect, count) => {
  const { width, height, rx, ry, x, y } = rect;
  const sideConnectorsCount = Math.round(count / 4);
  const sideSpacesCount = sideConnectorsCount + 1;
  const startX = x + rx;
  const startY = x + ry;

  const widthConnectorOffset = (width - rx * 2) / (sideSpacesCount);
  const heightConnectorOffset = (height - ry * 2) / (sideSpacesCount);

  const sideConnectors = range(1, sideConnectorsCount + 1);

  const topConnectors = sideConnectors.map(index => generateConnectorPoint(
    startX + widthConnectorOffset * index,
    y,
  ));

  const rightConnectors = sideConnectors.map(index => generateConnectorPoint(
    x + width,
    startY + heightConnectorOffset * index,
  ));

  const bottomConnectors = sideConnectors.map(index => generateConnectorPoint(
    startX + widthConnectorOffset * index,
    y + height,
  ));

  const leftConnectors = sideConnectors.map(index => generateConnectorPoint(
    x,
    startY + heightConnectorOffset * index,
  ));

  const connectors = [
    ...topConnectors,
    ...rightConnectors,
    ...bottomConnectors,
    ...leftConnectors,
  ];

  if (!rx && !ry) {
    /* Angles connectors */
    connectors.push(
      generateConnectorPoint(x, y),
      generateConnectorPoint(x + width, y),
      generateConnectorPoint(x + width, y + height),
      generateConnectorPoint(x, y + height),
    );
  }

  return connectors;
};
