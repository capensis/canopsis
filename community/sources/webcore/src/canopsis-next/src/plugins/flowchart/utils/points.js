/**
 * Calculate points with ghosts points
 *
 * @param {Array} points
 * @returns {Array}
 */
export const getPointsWithGhosts = points => points.reduce((acc, point, index) => {
  const nextIndex = index + 1;
  const nextPoint = points[nextIndex];

  acc.push({ ...point, index });

  if (nextPoint) {
    acc.push({
      x: (point.x + nextPoint.x) / 2,
      y: (point.y + nextPoint.y) / 2,
      ghost: true,
      index: nextIndex,
    });
  }

  return acc;
}, []);
