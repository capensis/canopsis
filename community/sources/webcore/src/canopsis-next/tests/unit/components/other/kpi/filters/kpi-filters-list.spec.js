import { generateRenderer } from '@unit/utils/vue';

import KpiFiltersList from '@/components/other/kpi/filters/kpi-filters-list.vue';
import CAdvancedDataTable from '@/components/common/table/c-advanced-data-table.vue';

const stubs = {
  'c-advanced-data-table': CAdvancedDataTable,
  'c-search-field': true,
  'c-expand-btn': true,
  'c-action-btn': true,
  'c-table-pagination': true,
  'kpi-filters-expand-item': true,
};

const selectExpandButtonByRow = (wrapper, index) => wrapper
  .findAll('tbody > tr')
  .at(index)
  .find('.v-data-table__expand-icon');

describe('kpi-filters-list', () => {
  const filtersItems = [
    {
      _id: 'c0ed9d92-67eb-4dc7-a2ab-9a551d45b9bf',
      name: 'Filter',
      created: 1614861888,
      updated: 1614861888,
    },
    {
      _id: '441a2a17-9036-48a3-9ff7-f393487395a9',
      name: 'Filter 2',
      created: 1614863990,
      updated: 1614863990,
    },
    {
      _id: '1cae4b8a-f598-480a-ad0c-b0a89a5c2e93',
      name: 'Filter 3',
      created: 1614864049,
      updated: 1614864049,
    },
    {
      _id: 'c46bffd9-8f5a-4c6c-b045-416e23ab1d44',
      name: 'Filter 4',
      created: 1614857014,
      updated: 1614857014,
    },
    {
      _id: 'd2403af7-712d-4353-911e-376f7a8053a7',
      name: 'Filter 5',
      created: 1613620731,
      updated: 1613620731,
    },
    {
      _id: '9bbb623c-7537-4c3b-afc0-0ace4f25a76b',
      name: 'Filter 6',
      created: 1613620721,
      updated: 1613620721,
    },
    {
      _id: 'fd35fcc4-36b0-445d-85be-999cc939047a',
      name: 'Filter 7',
      created: 1614864251,
      updated: 1614864251,
    },
    {
      _id: '70bfae47-cfdf-4a2c-9b43-3427f6aabea2',
      name: 'Filter 8',
      created: 1622781697,
      updated: 1622781697,
    },
    {
      _id: 'b3f67a16-019a-4694-9d74-ed762affaa04',
      name: 'Filter 9',
      created: 1615435601,
      updated: 1615435601,
    },
    {
      _id: 'e1f3e64a-dc99-42ed-af72-d8678f2e62bf',
      name: 'Filter 10',
      created: 1615442556,
      updated: 1615442556,
    },
    {
      _id: '15094f5a-9472-4700-b0cd-52305f754754',
      name: 'Filter 11',
      created: 1615440560,
      updated: 1615440560,
    },
  ];

  const snapshotFactory = generateRenderer(KpiFiltersList, { stubs });

  it('Renders `kpi-filters-list` with default props', () => {
    const wrapper = snapshotFactory({
      propsData: {
        filters: [],
        options: {},
      },
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('Renders `kpi-filters-list` with custom props', () => {
    const wrapper = snapshotFactory({
      propsData: {
        filters: filtersItems,
        options: {
          page: 2,
          itemsPerPage: 10,
          search: 'Filter',
          sortBy: ['created'],
          sortDesc: [],
        },
        totalItems: 50,
        pending: true,
        removable: true,
        updatable: true,
        duplicable: true,
      },
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('Renders `kpi-filters-list` with expanded panel', async () => {
    const wrapper = snapshotFactory({
      propsData: {
        filters: filtersItems,
        options: {
          page: 1,
          itemsPerPage: 10,
          search: '',
          sortBy: [],
          sortDesc: [],
        },
        totalItems: 50,
      },
    });

    await selectExpandButtonByRow(wrapper, 0).triggerCustomEvent('expand');

    expect(wrapper).toMatchSnapshot();
  });

  it('Renders `kpi-filters-list` with updatable and old_entity_patterns', async () => {
    const wrapper = snapshotFactory({
      propsData: {
        filters: filtersItems.map(item => ({ ...item, old_entity_patterns: true })),
        options: {
          page: 1,
          itemsPerPage: 10,
          search: '',
          sortBy: [],
          sortDesc: [],
        },
        updatable: true,
        totalItems: 50,
      },
    });

    expect(wrapper).toMatchSnapshot();
  });
});
