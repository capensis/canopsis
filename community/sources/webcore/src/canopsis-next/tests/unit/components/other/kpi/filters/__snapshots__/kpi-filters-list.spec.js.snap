// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kpi-filters-list Renders \`kpi-filters-list\` with custom props 1`] = `
<div
  class="c-advanced-data-table white white"
>
  <div
    class="layout row wrap"
  >
    <div
      class="flex xs4"
    >
      <c-search-field-stub
        value="Filter"
      />
    </div>
    <!---->
  </div>
  <div>
    <div
      class="v-table__overflow"
    >
      <table
        class="v-datatable v-table theme--light"
      >
        <thead>
          <tr>
            <th
              aria-label=": Not sorted."
              aria-sort="none"
              class="column text-xs-left"
              role="columnheader"
              scope="col"
              width="20"
            >
              
            </th>
            <th
              aria-label="Name: Not sorted. Activate to sort ascending."
              aria-sort="none"
              class="column sortable text-xs-left"
              role="columnheader"
              scope="col"
              tabindex="0"
            >
              Name
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
                style="font-size: 16px;"
              >
                arrow_upward
              </i>
            </th>
            <th
              aria-label="Creation date: Sorted descending. Activate to remove sorting."
              aria-sort="descending"
              class="column active desc sortable text-xs-left"
              role="columnheader"
              scope="col"
              tabindex="0"
            >
              Creation date
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
                style="font-size: 16px;"
              >
                arrow_upward
              </i>
            </th>
            <th
              aria-label="Last modified on: Not sorted. Activate to sort ascending."
              aria-sort="none"
              class="column sortable text-xs-left"
              role="columnheader"
              scope="col"
              tabindex="0"
            >
              Last modified on
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
                style="font-size: 16px;"
              >
                arrow_upward
              </i>
            </th>
            <th
              aria-label="Actions: Not sorted."
              aria-sort="none"
              class="column text-xs-left"
              role="columnheader"
              scope="col"
            >
              Actions
            </th>
          </tr>
          <tr
            class="v-datatable__progress"
          >
            <th
              class="column"
              colspan="5"
            >
              <div
                aria-valuemax="100"
                aria-valuemin="0"
                class="v-progress-linear"
                role="progressbar"
                style="height: 2px;"
              >
                <div
                  class="v-progress-linear__background primary"
                  style="height: 2px; opacity: 0.3; width: 100%;"
                />
                <div
                  class="v-progress-linear__bar"
                >
                  <transition-stub
                    name="fade-transition"
                  >
                    <div
                      class="v-progress-linear__bar__indeterminate v-progress-linear__bar__indeterminate--active"
                    >
                      <div
                        class="v-progress-linear__bar__indeterminate long primary"
                      />
                      <div
                        class="v-progress-linear__bar__indeterminate short primary"
                      />
                    </div>
                  </transition-stub>
                  <transition-stub
                    name="slide-x-transition"
                  />
                </div>
              </div>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter
            </td>
            <td>
              04/03/2021 13:44:48
            </td>
            <td>
              04/03/2021 13:44:48
            </td>
            <td>
              <div
                class="layout row"
              >
                <c-action-btn-stub
                  type="edit"
                />
                <c-action-btn-stub
                  type="duplicate"
                />
                <c-action-btn-stub
                  type="delete"
                />
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 2
            </td>
            <td>
              04/03/2021 14:19:50
            </td>
            <td>
              04/03/2021 14:19:50
            </td>
            <td>
              <div
                class="layout row"
              >
                <c-action-btn-stub
                  type="edit"
                />
                <c-action-btn-stub
                  type="duplicate"
                />
                <c-action-btn-stub
                  type="delete"
                />
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 3
            </td>
            <td>
              04/03/2021 14:20:49
            </td>
            <td>
              04/03/2021 14:20:49
            </td>
            <td>
              <div
                class="layout row"
              >
                <c-action-btn-stub
                  type="edit"
                />
                <c-action-btn-stub
                  type="duplicate"
                />
                <c-action-btn-stub
                  type="delete"
                />
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 4
            </td>
            <td>
              04/03/2021 12:23:34
            </td>
            <td>
              04/03/2021 12:23:34
            </td>
            <td>
              <div
                class="layout row"
              >
                <c-action-btn-stub
                  type="edit"
                />
                <c-action-btn-stub
                  type="duplicate"
                />
                <c-action-btn-stub
                  type="delete"
                />
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 5
            </td>
            <td>
              18/02/2021 4:58:51
            </td>
            <td>
              18/02/2021 4:58:51
            </td>
            <td>
              <div
                class="layout row"
              >
                <c-action-btn-stub
                  type="edit"
                />
                <c-action-btn-stub
                  type="duplicate"
                />
                <c-action-btn-stub
                  type="delete"
                />
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 6
            </td>
            <td>
              18/02/2021 4:58:41
            </td>
            <td>
              18/02/2021 4:58:41
            </td>
            <td>
              <div
                class="layout row"
              >
                <c-action-btn-stub
                  type="edit"
                />
                <c-action-btn-stub
                  type="duplicate"
                />
                <c-action-btn-stub
                  type="delete"
                />
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 7
            </td>
            <td>
              04/03/2021 14:24:11
            </td>
            <td>
              04/03/2021 14:24:11
            </td>
            <td>
              <div
                class="layout row"
              >
                <c-action-btn-stub
                  type="edit"
                />
                <c-action-btn-stub
                  type="duplicate"
                />
                <c-action-btn-stub
                  type="delete"
                />
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 8
            </td>
            <td>
              04/06/2021 6:41:37
            </td>
            <td>
              04/06/2021 6:41:37
            </td>
            <td>
              <div
                class="layout row"
              >
                <c-action-btn-stub
                  type="edit"
                />
                <c-action-btn-stub
                  type="duplicate"
                />
                <c-action-btn-stub
                  type="delete"
                />
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 9
            </td>
            <td>
              11/03/2021 5:06:41
            </td>
            <td>
              11/03/2021 5:06:41
            </td>
            <td>
              <div
                class="layout row"
              >
                <c-action-btn-stub
                  type="edit"
                />
                <c-action-btn-stub
                  type="duplicate"
                />
                <c-action-btn-stub
                  type="delete"
                />
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 10
            </td>
            <td>
              11/03/2021 7:02:36
            </td>
            <td>
              11/03/2021 7:02:36
            </td>
            <td>
              <div
                class="layout row"
              >
                <c-action-btn-stub
                  type="edit"
                />
                <c-action-btn-stub
                  type="duplicate"
                />
                <c-action-btn-stub
                  type="delete"
                />
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 11
            </td>
            <td>
              11/03/2021 6:29:20
            </td>
            <td>
              11/03/2021 6:29:20
            </td>
            <td>
              <div
                class="layout row"
              >
                <c-action-btn-stub
                  type="edit"
                />
                <c-action-btn-stub
                  type="duplicate"
                />
                <c-action-btn-stub
                  type="delete"
                />
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <c-table-pagination-stub
    page="2"
    rows-per-page="10"
    total-items="50"
  />
</div>
`;

exports[`kpi-filters-list Renders \`kpi-filters-list\` with default props 1`] = `
<div
  class="c-advanced-data-table white white"
>
  <div
    class="layout row wrap"
  >
    <div
      class="flex xs4"
    >
      <c-search-field-stub />
    </div>
    <!---->
  </div>
  <div>
    <div
      class="v-table__overflow"
    >
      <table
        class="v-datatable v-table theme--light"
      >
        <thead>
          <tr>
            <th
              aria-label=": Not sorted."
              aria-sort="none"
              class="column text-xs-left"
              role="columnheader"
              scope="col"
              width="20"
            >
              
            </th>
            <th
              aria-label="Name: Sorted ascending. Activate to sort descending."
              aria-sort="ascending"
              class="column active asc sortable text-xs-left"
              role="columnheader"
              scope="col"
              tabindex="0"
            >
              Name
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
                style="font-size: 16px;"
              >
                arrow_upward
              </i>
            </th>
            <th
              aria-label="Creation date: Not sorted. Activate to sort ascending."
              aria-sort="none"
              class="column sortable text-xs-left"
              role="columnheader"
              scope="col"
              tabindex="0"
            >
              Creation date
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
                style="font-size: 16px;"
              >
                arrow_upward
              </i>
            </th>
            <th
              aria-label="Last modified on: Not sorted. Activate to sort ascending."
              aria-sort="none"
              class="column sortable text-xs-left"
              role="columnheader"
              scope="col"
              tabindex="0"
            >
              Last modified on
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
                style="font-size: 16px;"
              >
                arrow_upward
              </i>
            </th>
            <th
              aria-label="Actions: Not sorted."
              aria-sort="none"
              class="column text-xs-left"
              role="columnheader"
              scope="col"
            >
              Actions
            </th>
          </tr>
          <tr
            class="v-datatable__progress"
          >
            <th
              class="column"
              colspan="5"
            />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="text-xs-center"
              colspan="5"
            >
              No data available
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <c-table-pagination-stub />
</div>
`;

exports[`kpi-filters-list Renders \`kpi-filters-list\` with expanded panel 1`] = `
<div
  class="c-advanced-data-table white white"
>
  <div
    class="layout row wrap"
  >
    <div
      class="flex xs4"
    >
      <c-search-field-stub
        value=""
      />
    </div>
    <!---->
  </div>
  <div>
    <div
      class="v-table__overflow"
    >
      <table
        class="v-datatable v-table theme--light"
      >
        <thead>
          <tr>
            <th
              aria-label=": Not sorted."
              aria-sort="none"
              class="column text-xs-left"
              role="columnheader"
              scope="col"
              width="20"
            >
              
            </th>
            <th
              aria-label="Name: Not sorted. Activate to sort ascending."
              aria-sort="none"
              class="column sortable text-xs-left"
              role="columnheader"
              scope="col"
              tabindex="0"
            >
              Name
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
                style="font-size: 16px;"
              >
                arrow_upward
              </i>
            </th>
            <th
              aria-label="Creation date: Not sorted. Activate to sort ascending."
              aria-sort="none"
              class="column sortable text-xs-left"
              role="columnheader"
              scope="col"
              tabindex="0"
            >
              Creation date
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
                style="font-size: 16px;"
              >
                arrow_upward
              </i>
            </th>
            <th
              aria-label="Last modified on: Not sorted. Activate to sort ascending."
              aria-sort="none"
              class="column sortable text-xs-left"
              role="columnheader"
              scope="col"
              tabindex="0"
            >
              Last modified on
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
                style="font-size: 16px;"
              >
                arrow_upward
              </i>
            </th>
            <th
              aria-label="Actions: Not sorted."
              aria-sort="none"
              class="column text-xs-left"
              role="columnheader"
              scope="col"
            >
              Actions
            </th>
          </tr>
          <tr
            class="v-datatable__progress"
          >
            <th
              class="column"
              colspan="5"
            />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                  expanded="true"
                />
              </div>
            </td>
            <td>
              Filter
            </td>
            <td>
              04/03/2021 13:44:48
            </td>
            <td>
              04/03/2021 13:44:48
            </td>
            <td>
              <div
                class="layout row"
              >
                <!---->
                <!---->
                <!---->
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            >
              <div
                class="v-datatable__expand-content"
              >
                <div
                  class="secondary lighten-2"
                >
                  <kpi-filters-expand-item-stub
                    filter="[object Object]"
                  />
                </div>
              </div>
            </transition-group-stub>
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 2
            </td>
            <td>
              04/03/2021 14:19:50
            </td>
            <td>
              04/03/2021 14:19:50
            </td>
            <td>
              <div
                class="layout row"
              >
                <!---->
                <!---->
                <!---->
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 3
            </td>
            <td>
              04/03/2021 14:20:49
            </td>
            <td>
              04/03/2021 14:20:49
            </td>
            <td>
              <div
                class="layout row"
              >
                <!---->
                <!---->
                <!---->
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 4
            </td>
            <td>
              04/03/2021 12:23:34
            </td>
            <td>
              04/03/2021 12:23:34
            </td>
            <td>
              <div
                class="layout row"
              >
                <!---->
                <!---->
                <!---->
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 5
            </td>
            <td>
              18/02/2021 4:58:51
            </td>
            <td>
              18/02/2021 4:58:51
            </td>
            <td>
              <div
                class="layout row"
              >
                <!---->
                <!---->
                <!---->
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 6
            </td>
            <td>
              18/02/2021 4:58:41
            </td>
            <td>
              18/02/2021 4:58:41
            </td>
            <td>
              <div
                class="layout row"
              >
                <!---->
                <!---->
                <!---->
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 7
            </td>
            <td>
              04/03/2021 14:24:11
            </td>
            <td>
              04/03/2021 14:24:11
            </td>
            <td>
              <div
                class="layout row"
              >
                <!---->
                <!---->
                <!---->
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 8
            </td>
            <td>
              04/06/2021 6:41:37
            </td>
            <td>
              04/06/2021 6:41:37
            </td>
            <td>
              <div
                class="layout row"
              >
                <!---->
                <!---->
                <!---->
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 9
            </td>
            <td>
              11/03/2021 5:06:41
            </td>
            <td>
              11/03/2021 5:06:41
            </td>
            <td>
              <div
                class="layout row"
              >
                <!---->
                <!---->
                <!---->
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 10
            </td>
            <td>
              11/03/2021 7:02:36
            </td>
            <td>
              11/03/2021 7:02:36
            </td>
            <td>
              <div
                class="layout row"
              >
                <!---->
                <!---->
                <!---->
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
          <tr>
            <td>
              <div
                class="layout c-checkbox-wrapper row justify-start"
              >
                <!---->
                <c-expand-btn-stub
                  class="ml-2"
                />
              </div>
            </td>
            <td>
              Filter 11
            </td>
            <td>
              11/03/2021 6:29:20
            </td>
            <td>
              11/03/2021 6:29:20
            </td>
            <td>
              <div
                class="layout row"
              >
                <!---->
                <!---->
                <!---->
              </div>
            </td>
          </tr>
          <tr
            class="v-datatable__expand-row"
          >
            <transition-group-stub
              class="v-datatable__expand-col"
              colspan="5"
              tag="td"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <c-table-pagination-stub
    page="1"
    rows-per-page="10"
    total-items="50"
  />
</div>
`;
